feature: repeater and builder item drag and drop ordering

jacksleight opened the pull request:
Drag and drop ordering for repeater items:

https://user-images.githubusercontent.com/126740/160114202-a1a2a39b-44e6-438b-a466-0bf419be9c4c.mov

jacksleight committed:
feature: Repeater item drag and drop

jacksleight committed:
Tweak

danharrin commented:
Thanks! Few points...
- Test failing
- I don't think we need extra button?
- Does this work with nested repeaters?
- How about the builder?

pxlrbt commented:
Maybe we should integrate our own Livewire integration of SortableJS. It's less than 100 LOC in this case and we don't need to rely on 3rd party packages. There is not much activity on that repo currently.

pxlrbt commented:
Btw tests are failing because you bumped min PHP version on 2.x 😅

jacksleight commented:
Yeah test error seems to be because of the composer requirements?

Works fine with nested repeaters.

Will change the drag handle as @pxlrbt suggested, and look at builder as well.

jacksleight committed:
New grab handle style

jacksleight committed:
Add builder support

jacksleight committed:
Switch dependency to sortablejs directly, integrate Livewire directive

jacksleight commented:
OK, made the suggested changes. I'm a bit unsure on how/where is best to import the custom Livewire directive, shout if it's wrong.

New drag handle style:

![Screenshot 2022-03-25 at 15 17 12](https://user-images.githubusercontent.com/126740/160149256-8b1f118b-b6e9-42d3-b28b-01fa36737966.png)

pxlrbt commented:
Not sure how to improve this, but the drag handle feels a bit off 🙈

jacksleight commented:
Mmmm yeah, I do feel it's a bit close to the field label.

pxlrbt commented:
> Edit: Also the Livewire directive is basically a copy paste from the other package, but could be customised now it's in Filament.

Wasn't the other implementation about 75 lines instead of 30? 🤔

jacksleight commented:
Oh, no they're self-contained at the moment, you can't move something from one nested to repeater to a parent/child repeater at the moment... Do the regular up/down buttons support that?

pxlrbt commented:
Sure, you're right. Maybe that's out of scope for now.

jacksleight commented:
Yeah, certainly be nice to have, but maybe one for another day 🙂 .

jacksleight commented:
Now we're using a custom Livewire directive we shouldn't need the extra dispatch event methods anymore. I'll refactor that.

wychoong commented:
maybe have the drag handle at the top center, heroicon-s-dots-horizontal, like ipad multitask dots?

panakour commented:
or maybe with a settings tooltip which let have more actions in the future, something like octobercms repeater
![Screenshot from 2022-03-26 13-14-56](https://user-images.githubusercontent.com/12927166/160236943-a7395cd1-84a7-49ba-8424-3f77ed074ae5.png)

jacksleight committed:
Refactor sortablejs/livewire integration

jacksleight committed:
Cleanup

jacksleight commented:
I've knocked up a few drag handle variations, 1 is the current version.

I'd personally lean towards 4, maybe 3, but I'm happy to implement whichever people prefer, or something else entirely. Let me know your thoughts.

![Screenshot 2022-03-26 at 16 04 22](https://user-images.githubusercontent.com/126740/160247998-7f020aec-cca6-4f87-991b-4386f21151c4.png)
![Screenshot 2022-03-26 at 16 03 45](https://user-images.githubusercontent.com/126740/160247999-a522f038-ea86-4d6d-9ea4-f985729a357c.png)
![Screenshot 2022-03-26 at 16 06 47](https://user-images.githubusercontent.com/126740/160247997-1eb7a389-db96-4f30-a65e-50ebef87d623.png)
![Screenshot 2022-03-26 at 16 07 41](https://user-images.githubusercontent.com/126740/160247996-4f09cab9-008e-47b9-b304-5df4e3a989e1.png)

panakour commented:
I also prefer the 4

martinmildner commented:
What about an up and down arrow?
<img width="47" alt="Screenshot-2022-03-27 at 14 25 53@2x" src="https://user-images.githubusercontent.com/789541/160281261-5db87c78-a125-40aa-84af-4d9cd8f2b2b1.png">

pxlrbt commented:
> What about an up and down arrow? <img alt="Screenshot-2022-03-27 at 14 25 53@2x" width="47" src="https://user-images.githubusercontent.com/789541/160281261-5db87c78-a125-40aa-84af-4d9cd8f2b2b1.png">

Check out the video in the initial PR comment ;)

jacksleight committed:
Tweak

danharrin commented:
I think 3 is my favourite, but we can get rid the old up / down buttons maybe? What do you think?

wychoong commented:
I feel 4 is crowded but of course I’m biased towards 3 😆 
If it’s not too hard to maintain, maybe ->showDragHandle(left/right/center) and ->hideSortButtons()

pxlrbt commented:
I guess with Drag & Drop we could remove the Up/Down arrow and get some space.

What about something like this? Inspired by other block editors and the October CMS solution. It takes up some more space but also gives a bigger dragging handle (whole grey background area)

1) Left
<img width="400" alt="Bildschirmfoto 2022-03-27 um 17 11 32" src="https://user-images.githubusercontent.com/22632550/160288121-4b1f4303-e77a-417c-9117-21b308833a9d.png">

2) Right
<img width="392" alt="Bildschirmfoto 2022-03-27 um 17 19 42" src="https://user-images.githubusercontent.com/22632550/160288437-a2081b55-537e-4100-8533-e44e6260edc6.png">

3) Top
<img width="395" alt="Bildschirmfoto 2022-03-27 um 17 30 20" src="https://user-images.githubusercontent.com/22632550/160288825-2ee0c042-6f07-49ba-b8b3-acbb0f9b3f9b.png">

martinmildner commented:
+1 for 5 (aligned top left)

jacksleight committed:
Update drag handle, remove arrows, add arrow key listeners

jacksleight commented:
I really like @panakour and @pxlrbt’s ideas of having a larger handle that could include the other actions as well. And possibly making this configurable as @wychoong suggested. 

But that probably needs more discussion, and I think it’d be good to get the core drag and drop functionality in place, so I’m gonna focus on that for this PR.

> I think 3 and 4 are my favourites, but we can get rid the old up / down buttons maybe? What do you think?

I’ve implemented 4 for now. That’s the simplest solution with the smallest UI impact.

I’ve removed the up/down buttons but added arrow up/down key listeners to the drag handle, so it’s still keyboard accessible. Livewire seems to get really glitchy if you add/remove these listeners on changes, so instead of toggling them in the view like the old buttons were I've added a check to the `array_move_after` and `array_move_before` functions.

jacksleight committed:
Icon pixel tweak

danharrin commented on packages/forms/src/helpers.php:
```diff
@@ -8,7 +8,10 @@ function array_move_after(array $array, $keyToMoveAfter): array
         $indexToMoveAfter = array_search($keyToMoveAfter, $keys);
-        $keyToMoveBefore = $keys[$indexToMoveAfter + 1];
+        $keyToMoveBefore = $keys[$indexToMoveAfter + 1] ?? null;
+        if (!isset($keyToMoveBefore)) {
```
```suggestion
        if (filled($keyToMoveBefore)) {
```

danharrin commented on packages/forms/src/helpers.php:
```diff
@@ -30,7 +33,10 @@ function array_move_before(array $array, $keyToMoveBefore): array
         $indexToMoveBefore = array_search($keyToMoveBefore, $keys);
-        $keyToMoveAfter = $keys[$indexToMoveBefore - 1];
+        $keyToMoveAfter = $keys[$indexToMoveBefore - 1] ?? null;
+        if (!isset($keyToMoveAfter)) {
```
```suggestion
        if (filled($keyToMoveAfter)) {
```

danharrin commented on packages/forms/src/helpers.php:
```diff
@@ -30,7 +33,10 @@ function array_move_before(array $array, $keyToMoveBefore): array
 
         $indexToMoveBefore = array_search($keyToMoveBefore, $keys);
-        $keyToMoveAfter = $keys[$indexToMoveBefore - 1];
+        $keyToMoveAfter = $keys[$indexToMoveBefore - 1] ?? null;
```
```suggestion
        $keyToMoveAfter = $keys[$indexToMoveBefore - 1] ?? null;
        
```

danharrin commented on packages/forms/src/helpers.php:
```diff
@@ -8,7 +8,10 @@ function array_move_after(array $array, $keyToMoveAfter): array
 
         $indexToMoveAfter = array_search($keyToMoveAfter, $keys);
-        $keyToMoveBefore = $keys[$indexToMoveAfter + 1];
+        $keyToMoveBefore = $keys[$indexToMoveAfter + 1] ?? null;
```
```suggestion
        $keyToMoveBefore = $keys[$indexToMoveAfter + 1] ?? null;
        
```

danharrin commented on packages/forms/src/Components/Repeater.php:
```diff
@@ -111,6 +111,26 @@ function (Repeater $component, string $statePath, string $uuidToMoveUp): void {
+                    data_set($livewire, $statePath, $items);
+                },
+            ],
+            'repeater::moveItems' => [
```
Since we add this new event listener for moving items, can we remove the old up/down button listeners for Repeater & Builder?

invaders-xx replied:
Maybe I am late in the conversation, but would it be possible to have the Drap&Drop placement driven by config. I don't know the complexity in implementing that

jacksleight replied:
The up/down listeners are still used by the arrow-up/arrow-down key events on the drag handle. Without those you wouldn't be able to move items via the keyboard.

jacksleight committed:
Update packages/forms/src/helpers.php

Co-authored-by: Dan Harrin <git@danharrin.com>

jacksleight committed:
Update packages/forms/src/helpers.php

Co-authored-by: Dan Harrin <git@danharrin.com>

jacksleight committed:
Update packages/forms/src/helpers.php

Co-authored-by: Dan Harrin <git@danharrin.com>

jacksleight committed:
Update packages/forms/src/helpers.php

Co-authored-by: Dan Harrin <git@danharrin.com>

a21ns1g4ts commented:
> I think 3 and 4 are my favourites, but we can get rid the old up / down buttons maybe? What do you think?

Yep. we can keep all clean

danharrin commented:
Thank you very much! 🍻

danharrin commented:
Hey @jacksleight, after multiple bug reports it seems that this breaks sorting for nested repeaters. I've opened a PR which removes this feature here - #2156.

I really don't want to merge it, but I've spent a while trying to find a solution and I cannot. It feels like Livewire is conflicting with the sortable package, which causes the moved items to appear above the parent repeater (DOM-diffing issue?). @ryangjchandler or @jacksleight if you're able to look into this today or tomorrow, that would be great. But if you don't have any time then that's absolutely fine, but I will need to merge #2156, probably tomorrow afternoon.

danharrin commented:
To replicate the issue:

```php
Repeater::make('options')
    ->schema([
        Repeater::make('values')
            ->schema([
                TextInput::make('value'),
            ])
    ]),
```

Add two items to the `values` repeater and then swap their positions.

danharrin commented:
Thanks @ryangjchandler, fixed by #2164.

jacksleight commented:
Hey, sorry I've been away so didn't catch this, wasn't ignoring it 😄 . Glad you found a fix!

jacksleight commented:
@danharrin Hey, sorry I've been away so didn't catch this, wasn't ignoring it 😄 . Glad you found a fix!

danharrin commented:
No worries mate, all sorted! Hope you had a good break
